README.md
# prodig-external-mail-sender [ALPHA]

A Node.js microservice for sending emails as an external task handed over by 
Camunda WFMS.

It is intended to be more flexible than the integrated mail-connector and
fits better into the microservice architecture we want to set up (less code in 
BPMN and Camunda). It still supports variables of the legacy connector.

The project is based on nodemailer, handlebars, mjml, camunda sdk and 
a support package ([prodig-external-deps](https://www.npmjs.com/package/@kai-raschke/prodig-external-deps)) - 
almost not micro any more :)

Email templates can be changed just-in-time without deploying a new process version
which we prefer in our use case.

It was built with PM2 as process manager in mind so all settings are set in the
app.json file in a [PM2 process file](http://pm2.keymetrics.io/docs/usage/application-declaration/) 
format.

For more information about our project, in which we use this service, you can visit our
webpage [Prodig@Students](https://prodig.uni-halle.de).

## Content

[Features](#features) \
[Setup](#setup) \
[Usage](#how-to-use) \
[How it works](#how-it-works) \
[BPMN examples](#bpmn-examples) \
[Email templates](#email-templates) \
[mailVariables object](#mailvariables-object) \
[Options](#options) \
[TODO](#todo) \
[License](#license)

## Features

- Connecting to Camunda External Task queue to process service tasks
- Sending emails with nodemailer
- Uses handlebars template engine to render Camunda task variables in emails
- Uses MJML templates (optional) to create newsletter like email templates
- Logging with bunyan (file-rotation) and optional Graylog backend

## Setup

1. Clone the repository 

    `git clone`

2. Install npm dependencies
    `npm install`

3. Rename `app.json.example` to `app.json`

4. Configure options in `app.json`

5. Run
    1. With nodejs `npm start`
    2. With PM2 `pm2 start app.json`


**Requirements:**
- No JAVA
- NodeJS >= 8.12.0 `tested and developed with, newer should work`
- Camunda >= 8.9 `prior 8.9 will not support async response`

## Usage

### Mock server

To send your first mail through the service you can use the mock server 
functionality to test your templates without Camunda running.

1. Set up the project and set the environment variable NODE_ENV to 
"development"
2. Start the app - `npm start`
3. Use Postman or curl to send a POST request to the mock server. Process
variables are submitted in POST body as JSON.

Example mock request:
```
curl -X POST http://localhost:45321 \
    -H 'Content-Type: application/json' \
    -d '{
  	"mailTemplate": "prodig.test",
  	"mailVariables": {
  		"to": "your.mail@example.net",
  	},
  	"mailOptions": {
  		"forceRealSend": "true"
  	}
  }'
```

### BPMN

1. Set up the project and configure a Camunda instance (BASE_URL)
2. Set the environment variable for topic subscription (TOPIC)
3. Set the external task topic in your BPMN 
4. Deploy your BPMN (... try one of the examples)
5. Start the app - `npm start`
6. Start the process through tasklist

You can use one of the examples to get started - [BPMN examples](#bpmn-examples)

## Environments

### Development

`NODE_ENV = "development"`
- Mock server is activated and can be used by a HTTP POST to localhost:45321
- Nodemailer streamed JSON transport is used as primary SMTP service
- SMTP settings are respected but not used (until forceRealSend is used as option)
- Full debugging

### Staging

`NODE_ENV = "staging"`
- Mock server will not be activated
- SMTP settings are respected but mail "to" is replaced with the email set in 
  env variable (MAIL_STAGING) or your "from" address (to prevent sending real mails within the staging environment)
- Less debugging (>= info)

### Production

`NODE_ENV = "production"`
- Mock server will not be activated
- SMTP settings are respected
- Less debugging (>= warning)

## How it works

The service uses the JS version of Camunda External Task Client. The library is 
acquired through the support package "prodig-external-deps" which is a base package
to all my microservices I use in our project.

The External Task Client uses Camundas task queue to process external tasks and 
responds to them. For more information see Camunda docs at [GitHub Project](https://github.com/camunda/camunda-external-task-client-js)

This service is intended to be run by PM2 Node.js process manager. The app.json 
is prepared to be used for starting the service. For more information 
about PM2 see [PM2 Homepage](http://pm2.keymetrics.io/).

Email templating can be done with MJML ([Homepage](https://mjml.io/)) which is an easy way to get well 
styled emails with less code.

All templates are rendered with handlebars to fill in Camunda process variables 
you hand over with your task.

Sending mails is done by nodemailer ([Homepage](https://nodemailer.com)) which uses basic SMTP settings.

Logging is done with bunyan.

## BPMN examples

The project includes BPMN examples to use with the service.

You can refer to the BPMN and example config in the project which uses 
'test.external.mail.sender' as topic subscription by default.

**Send single mail legacy**

File: send-single-mail-legacy.bpmn

Service task is used with Camunda mail-connector variables as input parameters.
You can use the parameters as described in the connector docs, including FreeMaker templates.

**Send single mail JSON**

File: send-single-mail.bpmn

Mail settings are saved in a process Variable (mailVariables) as stringified JSON.
Template is saved in mailTemplate.

**Send multiple mails**

File: send-multiple-mail.bpmn

Service task is used as sequential multiple instance task which sends all emails 
of a list one after another to the external task queue.

To generate a collection element a script task is utilized beforehand.
The list of emails could also be a process variable (e. g. set by an user task).

```
var ArrayList = Java.type("java.util.ArrayList")
var addressArrayList = new ArrayList()

//Semicolon separated list of emails
var list = "your.mail@example.net;your.mail.2@example.net"

var emails = list.split(';')

for (i in emails) {
    var email = emails[i]

    try {
        var addressObject = {
            to: email
        }

        var addressObjectStringified = JSON.stringify(addressObject)
        addressArrayList.add(addressObjectStringified)
    } catch (ex) {
        print(ex)
    }
}

execution.setVariable('emailList', addressArrayList)
```

## Email templates

Templating uses handlebars or handlebars/mjml to render emails.

All templates are organized in a special folder structure in `emails` and separated
by purpose (subject, text, html).

An existing mjml template file is prioritized over the handlebars file.

**Structure**

To create a new template you need to create a new subfolder under `emails`.
Name of the folder is later used as template name.

Rendering is done by `email-templates` which uses 3 files:

- subject - Email subject
- text - Email in text only format
- html - Email in html format

To be recognized by the handlebars rendering engine, the files extension is
.handlebars.

The folder structure looks like this:

```
emails -\
        \- templateName -\
                         |- subject.handlebars
                         |- text.handlebars
                         \- html.handlebars
```

**Handlebars** is always used to render process variables contained in the 
mailVariables object.

A handlebars expression is composed as `{{` variable `}}`, for example:
```
Hello {{fullname}},

you won the second price of our big holiday lottery.

Regards, Sp4mking2000
```

For all features of handlebars see their own docs at [https://handlebarsjs.com](https://handlebarsjs.com/).

**MJML** can be used to style templates. It is a responsive email framework
making it easy to create well styled emails.

To see all features see their own docs at [https://mjml.io](https://mjml.io/).

## mailVariables object

The mail variables object is used to send a stringified JSON object as process
variable to the service.

It was introduced to make the sequential multiple instance task working for
multiple receivers and is inspired by the original variable of mail-connector.

- **to**: Address of receiver
- **cc**: Carbon copy receiver
- **bcc**: Blind carbon copy receiver
- **alias**: Send-as alias
- **replyTo**: Reply to address if receiver wants to answer your mail
- **anyRandomKey**: Process variable to render in templates

Example:

```
{
    to:     "receiver@example.net",
    cc:     "cc@example.net",
    bcc:    "bcc@example.net",
    alias:  "My name is",
    replyTo:"inbox@example.net",
    name:   "John Jackson"
}
```

## Options

All application settings are set through [PM2 process file](http://pm2.keymetrics.io/docs/usage/application-declaration/) 
format. You can find an example in app.json.example.

Settings are read from app.json.

Settings are string values (e. g. "true" for bool) and will be parsed if necessary.

| ENV | Type | Default | Required | Description |
| --- | ---- | --- | --- | --- |
| NODE_ENV | development / staging / production | - | X | NodeJS execution environment |
| MAX_TASK | {Number} | 10 | | See [Camunda docs](https://github.com/camunda/camunda-external-task-client-js/blob/master/docs/Client.md#new-clientoptions) |
| INTERVAL | {Number} | 300 | | See [Camunda docs](https://github.com/camunda/camunda-external-task-client-js/blob/master/docs/Client.md#new-clientoptions) |
| ASYNC_RESPONSE_TIMEOUT | {Number} | - | | See [Camunda docs](https://github.com/camunda/camunda-external-task-client-js/blob/master/docs/Client.md#new-clientoptions) |
| LOCK_DURATION | {Number} | 50000 | | See [Camunda docs](https://github.com/camunda/camunda-external-task-client-js/blob/master/docs/Client.md#new-clientoptions) |
| BASE_URL | {String} | - | | See [Camunda docs](https://github.com/camunda/camunda-external-task-client-js/blob/master/docs/Client.md#new-clientoptions) |
| AUTH | {Boolean} | false | | Activate basic auth for Camunda Rest API |
| USER | {String} | - | If AUTH == true | Basic auth user |
| PASS | {String} | - | If AUTH == true | Basic auth password |
| SMTP_SERVER | {String} | localhost | | See "host" in [Nodemailer docs](https://nodemailer.com/smtp/) |
| SMTP_PORT | {Number} | 587 / 465 | | See "port" in [Nodemailer docs](https://nodemailer.com/smtp/) |
| SMTP_SECURE | {Boolean} | false | | See "secure" in [Nodemailer docs](https://nodemailer.com/smtp/) |
| SMTP_REQUIRE_TLS | {Boolean} | false | | See "requireTLS" in [Nodemailer docs](https://nodemailer.com/smtp/) |
| SMTP_AUTH | {Boolean} | - | | Activates SMTP auth for nodemailer |
| SMTP_USER | {String} | - | If SMTP_AUTH == true | See "user" in [Nodemailer docs](https://nodemailer.com/smtp/#authentication) |
| SMTP_PASS | {String} | - | If SMTP_AUTH == true | See "pass" in [Nodemailer docs](https://nodemailer.com/smtp/#authentication) |
| MAIL_FROM | {String} | - | X | Mail address to send from |
| MAIL_STAGING | {String} | MAIL_FROM | | Overwrites 'to' address in staging mode |
| TOPIC | {String} | - | | Camunda External Task topic |
| LOG_CAMUNDA | {Boolean} | - | | Use Camunda External Client log |
| LOG_FILE | {Boolean} | - | | Use bunyan file/console log |

## TODO

- Send mails with attachment
- Use .env file if no app.json is available
- Validate nodemailer response
- Enable more customized logging
- Test useful cases (refactor current test and simplify execution)
- Production testing

## License

MIT

## Credits

Kai Raschke, Johannes Damarowsky

CHANGELOG.md
# Changelog

## 0.3.0 - 12/18/2018

- Enable/disable logging through env variables with new external-deps package
- Test deployment and email sending with mock mail server and `run-camunda` package

## 0.2.2 - 12/10/2018

- Added simple test case with smtp server

## 0.2.1 - 12/6/2018

- Update readme

## 0.2.0 - 12/4/2018

- Refactoring
- BPMN examples

## 0.1.2 - 10/30/2018

- Version push extra-dependency
- mjml renderer
- more templates (changed to mjml)
- graylog metric for 'Email-sent'

## 0.1.1 - 8/16/2018

- Support package refactored and mail.sender refactored
- Mainly uses app.json as settings file
- Supports more nodemailer settings

## 0.0.1 - 8/14/2018

- Initial release

test/test.js
/*
TODO needs refactoring and cleanup
 */
var assert = require('assert'); //replace with shouldjs
var should = require('should');
var fs = require('fs');

let client, log, gray,emailRenderer;
const { Variables } = require('camunda-external-task-client-js');
const nodemailer = require('nodemailer'),
    emailTemplates = require('email-templates');
const SMTPServer = require('smtp-server').SMTPServer;
const request = require('request');

let Client, serverOptions, voidFn = function () {};
describe('Array', function() {
    before(function(){
        process.env.NODE_ENV    = 'development';
        process.env.TOPIC       = 'test.external.mail.sender';
        process.env.BASE_URL    = "http://localhost:8080/engine-rest";
        process.env.MAX_TASK    = 1;
        process.env.MAX_TASK    = 1000;
        process.env.ASYNC_RESPONSE_TIMEOUT    = 10000;
        process.env.LOG_FILE = 'false';

        let deps = require('@kai-raschke/prodig-external-deps')();
        client = deps.client;
        log = deps.log;
        gray = deps.gray;
        emailRenderer = require('./../email-renderer');

        Client = client;

        serverOptions = {
            name: 'localhost',
            port: 2525,
            secure: true,
            authMethods: ['LOGIN'],
            onData: (stream, session, callback) => {
                const chunks = [];
                stream.on("data", function (chunk) {
                    chunks.push(chunk);
                });
                stream.on('end', function () {
                    let s = Buffer.concat(chunks).toString();
                    callback(null, s);
                });
            },
            onAuth: (auth, session, callback) => {
                //Allow all logins
                callback(null, {user: true});
            }
        };

        Client.subscribe(process.env.TOPIC, async function(){
            let transporter = nodemailer.createTransport({
                host: serverOptions.name,
                port: serverOptions.port,
                secure: serverOptions.secure,
                authMethod: serverOptions.authMethods[0], //'LOGIN'
                auth: { user: 'username', pass: 'pass' }, //not relevant, every auth is valid
                tls: { rejectUnauthorized: false } //self signed cert for localhost
            });

            let mail = await transporter.sendMail({
                from:   'from@example.com',
                to:     'to@example.com',
                text:   'My mail'
            });

            /* console.log(mail);
            { accepted: [ 'to@example.com' ],
              rejected: [],
              envelopeTime: 3,
              messageTime: 3,
              messageSize: 239,
              response: '250 Content-Type: text/plain',
              envelope: { from: 'from@example.com', to: [ 'to@example.com' ] },
              messageId: '<a3e70458-8e1e-af54-9253-3ae88251a19b@example.com>' }
             */
            //assert.strictEqual(err, null);
            //assert.notStrictEqual(info, null);
        })
    });

    describe('Test transport', function(){
        it('Should send 1 message', function(done){
            let transporter = nodemailer.createTransport({
                host: serverOptions.name,
                port: serverOptions.port,
                secure: serverOptions.secure,
                authMethod: serverOptions.authMethods[0], //'LOGIN'
                auth: { user: 'username', pass: 'pass' }, //not relevant, every auth is valid
                tls: { rejectUnauthorized: false } //self signed cert for localhost
            });

            const server = new SMTPServer(serverOptions);
            server.listen(serverOptions.port);

            transporter.sendMail({
                from:   'from@example.com',
                to:     'to@example.com',
                text:   'My mail'
            }, function(err, info) {
                //console.log(err, info);
                assert.strictEqual(err, null);
                assert.notStrictEqual(info, null);
                server.close();
                done();
            });
        });
    });

    describe('Test templates', function(){
        it('Should render test template', async function(){
            let emailRender = new emailTemplates({
                render: emailRenderer.render
            });

            let template = await emailRender.renderAll('prodig.test', { to: 'test@example.com' });

            should.equal(template.subject, 'Test Mail');
            should.equal(template.text, 'Test text to test@example.com');
            //should equal html version
        });
    });

    describe('Test Camunda', function(){
        it('Should deploy and send', function(done){
            Client.start();

            serverOptions.onData = (stream, session, callback) => {
                const chunks = [];
                stream.on("data", function (chunk) {
                    chunks.push(chunk);
                });
                stream.on('end', function () {
                    let s = Buffer.concat(chunks).toString();
                    Client.stop();
                    //server.close();
                    done();
                    callback(null, s);
                });
            };
            const server = new SMTPServer(serverOptions);
            server.listen(serverOptions.port);

            request.post({
                    url:'http://localhost:8080/engine-rest/deployment/create',
                    formData: {
                        'deployment-name': 'Send single mail',
                        'file': fs.createReadStream(__dirname + '/../bpmn/send-single-test-mail.bpmn')
                    }
                },
                function optionalCallback(err, httpResponse, body) {
                    if (err)
                        return console.error('upload failed:', err);

                    try {
                        request.post({
                                url: `http://localhost:8080/engine-rest/process-definition/key/send-single-test-mail/start`,
                                body: {},
                                json: true,
                            });
                        //done() is called in smtp server response
                    } catch (e) { throw e.response ? e.response.body : e; }
            });
        });
    });
});

mock.js
/**
 * Mock module to test mail templates
 * Module is required in development/test mode
 * Test can be done by sending http post to the local server (port 45321)
 * Post body must be application/json and variables "mailTemplate" and "mailVariables" set
 */

const http = require('http');
const port = 45321;
let senderFn;

//mock classes for testing
class _task{
    constructor(){
        this.variables = new _variables();
    }
}
class _variables{
    constructor(){
        this.mailTemplate = '';
        this.mailVariables = '';
        this.mailOptions = '';
    }

    get(name){
        return this[name];
    }

    set(name, value){
        this[name] = value;
    }
}
class _taskService{
    constructor(){}
    async complete(t, vars){
        console.log(t);
        console.log(vars);
    }
    async handleFailure(t, err){
        console.error(t);
        console.error(err);
    }
}

function start(main){
    senderFn = main;
    const server = http.createServer(requestHandler);
    server.listen(port, (err) => {
        if (err) {
            return console.error('Error: ', err)
        }

        console.log(`mock server is listening on ${port}`)
    });
}

const requestHandler = (request, response) => {
    const { headers, method, url } = request;
    if (request.method === 'POST') {
        let body = '';
        request.on('data', chunk => {
            body += chunk.toString();
        });
        request.on('end', () => {
            let content = JSON.parse(body);
            console.log(content);

            if(content){
                let t = new _task();

                t.variables.set('mailTemplate', content.mailTemplate);
                t.variables.set('mailVariables', JSON.stringify(content.mailVariables));
                t.variables.set('mailOptions', JSON.stringify((content.mailOptions || {})));

                let service = new _taskService();
                senderFn({task: t, taskService: service});
            }
            response.end('ok');
        });
    }
};

module.exports = { start };

email-renderer.js
const fs = require('fs'), path = require('path');
const { log } = require('@kai-raschke/prodig-external-deps')();
const mjml2html = require('mjml'), //.default;
    handlebars = require('handlebars');

function render (view, locals) {
    return new Promise((res, rej) => {
        log.debug(`Try view ${view}`);
        let pMjml = path.join(__dirname, 'emails', `${view}.handlebars.mjml`);
        let pHandlebars = path.join(__dirname, 'emails', `${view}.handlebars`);

        log.debug(`Try ${pMjml} and ${pHandlebars}`);
        log.debug(`Mjml format ${(fs.existsSync(pMjml) ? 'exists' : 'does not exist')}`);

        let templateFile, mode = "none";
        if(fs.existsSync(pMjml)){
            templateFile = pMjml;
            mode = "mjml";
        }
        else if(fs.existsSync(pHandlebars)){
            templateFile = pHandlebars;
            mode = "handlebars";
        }

        if(mode === "none"){
            res('');
        }
        else{
            fs.readFile(templateFile, async function read(err, data) {
                if (err) {
                    rej (err);
                }

                let template;
                //Render mjml if available
                if(mode === "mjml"){
                    let htmlOutput = mjml2html(data.toString(), {
                        //minify: true
                    });

                    template = handlebars.compile(htmlOutput.html);
                }
                else{
                    template = handlebars.compile(data.toString());
                }

                const result = template(locals);
                res(result);
            });
        }
    });
}

module.exports = { render };

index.js
'use strict';

const { Variables } = require('camunda-external-task-client-js');
const nodemailer = require('nodemailer'),
    emailTemplates = require('email-templates'),
    emailRenderer = require('./email-renderer');
const { client, log, gray } = require('@kai-raschke/prodig-external-deps')(
    /* Optional config object */
);

let transporter, transportForce;

/***
 * Main task for subscription
 * @param task
 * @param taskService
 * @returns {Promise<void>}
 */
let main = async function({ task, taskService }){
    try{
        const processVariables  = new Variables();

        //process.env variables are prioritized over JSON mail variables
        //JSON mail variables are prioritized over mail-connector variables

        //Variables like original camunda-mail connector https://github.com/camunda/camunda-bpm-mail#send-mails
        //including new replyTo
        let to                  = task.variables.get("to")              || undefined;
        let cc                  = task.variables.get("cc")              || undefined;
        let bcc                 = task.variables.get("bcc")             || undefined;
        let replyTo             = task.variables.get("replyTo")         || undefined;
        let fromAlias           = task.variables.get("fromAlias")       || undefined;
        let subject             = task.variables.get("subject")         || undefined;
        let text                = task.variables.get("text")            || undefined;
        let html                = task.variables.get("html")            || undefined;

        let mailTemplate        = task.variables.get("mailTemplate")    || undefined;
        let mailVariables       = task.variables.get("mailVariables")   || "{}";
        let mailOptions         = task.variables.get("mailOptions")     || "{}";

        mailVariables   = JSON.parse(mailVariables);
        mailOptions     = JSON.parse(mailOptions);

        mailOptions.forceRealSend   = (process.env.FORCE_REAL_SEND === 'true' ? true : mailOptions.forceRealSend || false); //Force smtp transporter in dev mode
        mailOptions.returnMessageId = mailOptions.returnMessageId || false; //return the Message ID of sent mail

        //Forces real smtp transporter when in dev mode
        if(mailOptions.forceRealSend){
            transporter = transportForce;
        }

        //If staging mode, overwrite receiver with static testing address, if not set, send to yourself
        const MAIL_STAGING = (process.env.MAIL_STAGING ? process.env.MAIL_STAGING : from); //In staging no real mail shall be sent
        const MAIL_FROM_ALIAS = process.env.MAIL_FROM_ALIAS || mailVariables.alias || fromAlias; //by priorities

        let envelope = {
            to: (process.env.NODE_ENV === 'staging' ? MAIL_STAGING : mailVariables.to || to),
            from: (MAIL_FROM_ALIAS ? { //Use Alias if there was one delivered
                name: MAIL_FROM_ALIAS,
                address: process.env.MAIL_FROM
            } : process.env.MAIL_FROM), //Otherwise use just from address
            cc:     mailVariables.cc        || cc,
            bcc:    mailVariables.bcc       || bcc,
            replyTo:mailVariables.replyTo   || replyTo,
        };

        let template = {
            subject, text, html
        };

        //If email template is defined render with engine otherwise use camunda-mail connector settings
        if(mailTemplate){
            try{
                let emailRender = new emailTemplates({
                    render: emailRenderer.render
                });

                template = await emailRender.renderAll(mailTemplate, mailVariables);
                log.debug(`Render templates done`);
            } catch(ex){
                await taskService.handleFailure(task, "An error occured while rendering the templates");
            }
        }
        else{
            template.subject = subject;
            template.text = text;
            template.html = html;
        }

        try{
            let mail = await transporter.sendMail({
                from:   envelope.from,
                to:     envelope.to,
                cc:     envelope.cc,
                bcc:    envelope.bcc,
                replyTo:envelope.replyTo,
                subject:template.subject,
                text:   template.text,
                html:   template.html
            });

            if(mail.accepted){
                if(mail.accepted.length > 0){
                    gray.warning('Email-sent', {type: 'metric', value: 1}); //graylog metrics
                }
                else{
                    log.warn('No mail was sent');
                }
            }

            if(mailOptions.returnMessageId === true){
                processVariables.set("messageId", mail.messageId);
            }

            //TODO check response for '250 ok' and/or 'accepted' matches the input addresses
            //TODO Throw BPMN errors on specific failure

            await taskService.complete(task, processVariables);
        } catch(ex){
            await taskService.handleFailure(task, "An error occured while sending the mail");
        }
    }
    catch(ex){
        log.error(ex);
        await taskService.handleFailure(task, "An error occured - strange ...");
    }
};

/***
 * Anonymous start function, subscribes to topic by given NODE_ENV
 */
(function start(){
    try {
        //nodemail smtp settings
        let smtpWorkflowConfig = {
            host: process.env.SMTP_SERVER, //defaults to localhost
            port: (isNaN(parseInt(process.env.SMTP_PORT)) ? undefined : parseInt(process.env.SMTP_PORT)), //defaults by nodemailer to 587/465 if not set
            secure: (process.env.SMTP_SECURE === 'true'), // upgrade later with STARTTLS
            requireTLS: (process.env.SMTP_REQUIRE_TLS === 'true'),
            logger: log,
            auth: (process.env.SMTP_AUTH === "true" ?
                    {
                        user: process.env.SMTP_USER,
                        pass: process.env.SMTP_PASS
                    } :
                    undefined
            )
        };

        transportForce = transporter = nodemailer.createTransport(smtpWorkflowConfig);
        // verify connection configuration
        transporter.verify(function(error, success) {
            if (error) {
                log.error(error);
            } else {
                log.info('Server is ready to take our messages');
            }
        });

        if(process.env.NODE_ENV === 'development' || process.env.NODE_ENV === 'test'){
            //Create JSON stream transporter when in dev mode
            transporter = nodemailer.createTransport({
                jsonTransport: true,
                logger: log
            });

            //allow testing Camunda external task through http interface if in dev mode
            require('./mock').start(main);
        }

        client.subscribe(process.env.TOPIC, main);
        client.start();
    } catch (e) {
        log.error(e);
    }
})();

